### PR TITLE
Remove PickledQuote and TastyString from library

### PR DESCRIPTION
**Pull request by nicolasstucki**
Remove PickledQuote and TastyString from the library and encode the pickled quotes with an efficient interface and code generation.

```scala
def unpickleExpr[T](pickled: String | List[String], typeHole: (Int, Seq[Any]) => Type[?], termHole: (Int, Seq[Any], QuoteContext) => Expr[?]): scala.quoted.Expr[T]
def unpickleType[T <: AnyKind](pickled: String | List[String], typeHole: (Int, Seq[Any]) => Type[?], termHole: (Int, Seq[Any], QuoteContext) => Expr[?]): scala.quoted.Type[T]

```

* `pickled` can take a `String` or `List[String]`s if it needs more than one string. In practice, we never have more than one string.
* Split `typeHole`/`termHole` to type them more precisely
* Uncurry arguments of `typeHole`/`termHole` to generate fewer closures. This also helps with beta-reduction of closures with the generated code of `typeHole`/`termHole` at each call site.

Part of #10222

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
For the following quote
```scala
'{ val y: T = $x; y }
```

We used to generate
```scala
x$2.asInstanceOf[scala.internal.quoted.QuoteContextInternal].unpickleExpr[
        evidence$1$_$1
      ](
        scala.internal.quoted.PickledQuote.make(
          scala.collection.immutable.Nil.::[String](
            
              "XKGrH5iAAJe0vOuFlwAAl2DupbEpAAGeAYRBU1RzAYF5AYEkAYxldmlkZW5jZSQxJF8Kg4KBgwGDQW55AYVzY2FsYQGMcXVvdGVUeXBlVGFnAYtDb21waWxlVGltZReBiAGIaW50ZXJuYWwCgoaKAYZxdW90ZWQCgouMAYY8aW5pdD4Cgo2IF4GPAoKQhz+CjpEBg0ZvbwGHPGVtcHR5PgGJUG9zaXRpb25zAYlGb28uc2NhbGGAwZO/jLaMkoqENIw1mIGKgTOKk4X/g4IzioOghKSM/4WAdYU2hv+DgTOgF62OdYdQdYk2jYiFcJJVM6xvk3WTUDeUlaKGAOcA+oSWA9B98Ki4fsP6hJ+H+4CXhIWAkYB/r4eGgJSB"
            
          )
        , 
          scala.collection.immutable.Nil.::[Seq[Any] => Any](
            {
              def $anonfun$3(x$0: Seq[Any]): (x$2.Nested) ?=> quoted.Expr[T] = 
                {
                  def $anonfun(using evidence$2: x$2.Nested): quoted.Expr[T] = x
                  closure($anonfun)
                }
              closure($anonfun$3)
            }
          ).::[Seq[Any] => Any](
            {
              def $anonfun$2(x$0: Seq[Any]): quoted.Type[T] = evidence$1
              closure($anonfun$2)
            }
          ).::[Seq[Any] => Any](
            {
              def $anonfun$1(x$0: Seq[Any]): quoted.Type[T] = evidence$1
              closure($anonfun$1)
            }
          )
        )
      )
```

Now we generate
```scala
x$2.asInstanceOf[scala.internal.quoted.QuoteContextInternal].unpickleExpr[
        evidence$1$_$1
      ](
        
          "XKGrH5iAAJe0vOuFlwAAl2DupbEpAAGeAYRBU1RzAYF5AYEkAYxldmlkZW5jZSQxJF8Kg4KBgwGDQW55AYVzY2FsYQGMcXVvdGVUeXBlVGFnAYtDb21waWxlVGltZReBiAGIaW50ZXJuYWwCgoaKAYZxdW90ZWQCgouMAYY8aW5pdD4Cgo2IF4GPAoKQhz+CjpEBg0ZvbwGHPGVtcHR5PgGJUG9zaXRpb25zAYlGb28uc2NhbGGAwZO/jLaMkoqENIw1mIGKgTOKk4X/g4IzioOghKSM/4WAdYU2hv+DgTOgF62OdYdQdYk2jYiFcJJVM6xvk3WTUDeUlaKGAOcA+oSWA9B98Ki4fsP6hJ+H+4CXhIWAkYB/r4eGgJSB"
        
      , 
        {
          def $anonfun(x$0: Int, x$1: Seq[Any]): scala.quoted.Type[<?>] = 
            x$0:(x$0 : Int) @unchecked match 
              {
                case 0 => evidence$1
                case 1 => evidence$1 // TODO optimization oporunity (deduplicate arguments)
              }
          closure($anonfun)
        }
      , 
        {
          def $anonfun(x$0: Int, x$1: Seq[Any], x$2: scala.quoted.QuoteContext)
            : 
          scala.quoted.Expr[Any] = 
            x$0:(x$0 : Int) @unchecked match 
              {
                case 2 => 
                  val evidence$2: x$2.Nested = x$2.asInstanceOf[x$2.Nested]
                  x
              }
          closure($anonfun)
        }
      )
```
We removed the creation of all the `::` list objects and only instantiate two functions (0 if there are no splices, 1 if there are only type or expr splices).

**Comment by liufengyun** [comment]
test performance with #quotes please

**Comment by liufengyun** [comment]
@nicolasstucki Need to rebase in order for benchmark test to succeed.

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Review by liufengyun — APPROVED** [review]
LGTM 👍

File: compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala [hunk 1]
```diff
@@ -194,23 +198,66 @@ class PickleQuotes extends MacroTransform {
...
+                  CaseDef(Literal(Constant(idx)), EmptyTree, splice.select(nme.apply).appliedTo(args(1)))
+                }
+                Match(args(0).annotated(New(ref(defn.UncheckedAnnot.typeRef))), cases)
+              }
+            )
+
+        // This and all closures in termSplices are removed by the BetaReduce phase
+        val termHoles = termSplices match
+          case Nil => Literal(Constant(null)) // keep pickled quote without splices as small as possible
+          case (firstSplice, _) :: _ =>
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala]: Replaced with an `if`

**Pull request merged**